### ISSUE TITLE
TypeError : 'Component' cannot be used as JSX component 

### ISSUE THREAD

**Issue by saurabhmehta1601** [vercel/next.js#37421]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Wed Aug 25 23:20:18 UTC 2021
    Binaries:
      Node: 14.18.2
      npm: 8.3.0
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.1.6
      react: 17.0.2
      react-dom: 17.0.2

### Which example does this report relate to?

with-redux

### Describe the Bug

I am using nextjs example [with-redux](https://github.com/vercel/next.js/tree/canary/examples/with-redux) . The current version of devDependencies `@types/react` and `@types/react-dom` gives type error ` Type Error 'Component' cannot be used as a JSX component.`  
Below is screenshot showing error 

![Screenshot (19)](https://user-images.githubusercontent.com/57309205/171814470-dffd6fc4-46e1-4eb1-a55b-2b801f001d0f.png)

The problem can be solved by upgrading the `@types/react` and `@types/react-dom`  dependencies

### Expected Behavior

TypeError should not occur 

### To Reproduce

If you do not see the typeError after cloning this example repository run next build to see typeError while build process .

**Comment by saurabhmehta1601** [comment]
Upgrading these depencencies solves the issue :
 
```
    "@types/react": "^18.0.10",
    "@types/react-dom": "^18.0.5"
```

**Comment by codemile** [comment]
This appears to be a bug with yarn. 

If you switch to npm and run `npm install` then it seems to be working fine for me.

I don't recommend using TypeScript defs from React 18 when you're running React 17.

**Comment by saurabhmehta1601** [comment]
> This appears to be a bug with yarn.
> 
> If you switch to npm and run `npm install` then it seems to be working fine for me.
> 
> I don't recommend using TypeScript defs from React 18 when you're running React 17.

What do you say about upgrading the react dependency ^18 . ??

**Comment by saurabhmehta1601** [comment]
> This appears to be a bug with yarn.
> 
> If you switch to npm and run `npm install` then it seems to be working fine for me.
> 
> I don't recommend using TypeScript defs from React 18 when you're running React 17.

I think so , but at least the build is not failing ( with yarn if it works just fine with npm ) .

**Comment by icyJoseph** [comment]
I've manage to trigger this by just:

```
yarn create next-app type-test --typescript
```

And then:

```
yarn add @types/react@17.0.45 @types/react-dom@17.0.11 --dev
```

And finally:

```
yarn add react@17.0.2 react-dom@17.0.2 --exact
```

Then you can't even build because of mismatched types.

**Comment by codemile** [comment]
This workaround is to add a resolution section to your `package.json` to ensure the correct TS defs were installed.

```json
"resolutions": {
    "**/@types/react": "^17.0.21",
    "**/@types/react-dom": "^17.0.9"
}
```

You can change the above versions to which ever ones you're using on your project.

I haven't fully tested the above, but it seems to have worked for me on a quick test.

> Note: I had to delete my node_modules and reinstall for this to work. Yarn doesn't seem to pick up on resolution changes.

**Comment by codemile** [comment]
As far as opening a PR. I'm not sure that's wise to apply this to all the examples. When NextJs is released with React 18 support, then I suspect this issue will go away. We would have to apply this fix to every example, and then remove it.

But... this problem is backwards breaking. Older versions of NextJs are going to have this issue, and it seems a dependency is getting installed with an updated change for React 18. I could still reproduce with `12.0.x` versions of NextJs.

I believe this problem was introduced by a deeper dependency, but I'm not sure which one is the cause. I've found people complaining of this problem in other libraries, because React 18 introduced breaking changes for the TypeScript defs, but people are using "^17.0.0 | ^18.0.0" for their peer dependencies. As an example, `react-router-dom` appears to be breaking builds for people still using React 17.

This is because React 18 is API compatible with React 17. So people make the peer dependency broad, but the TypeScript diffs are not backward compatible.

Give this a few months, because this is going to grow into a wider issue in the community. People are going to get upset.

**Comment by MingrenChen** [comment]
Having the same issue on an M1 MacOS. The build command is fine but vscode complain about 
```
Type '() => JSX.Element' is not assignable to type 'NextPage<{}, {}>'.
```
and
```
'Component' cannot be used as a JSX component.
```

**Comment by codemile** [comment]
@MingrenChen no, you're not having the same issue. You've installed types for React 18 with dependencies using React 17.  Use the proper types with Next 12.1.6. Which is React 17.
